python/mirascope/llm/providers/anthropic/beta/_utils/decode.py

```diff
@@ -0,0 +1,255 @@
+"""Beta Anthropic response decoding."""
+
+import json
+from typing import Any, TypeAlias, cast
+
+from anthropic.lib.streaming._beta_messages import (
+    BetaAsyncMessageStreamManager,
+    BetaMessageStreamManager,
+)
+from anthropic.types.beta import (
+    BetaContentBlock,
+    BetaRawMessageStreamEvent,
+    BetaRedactedThinkingBlockParam,
+    BetaTextBlockParam,
+    BetaThinkingBlockParam,
+    BetaToolUseBlockParam,
+)
+from anthropic.types.beta.parsed_beta_message import ParsedBetaMessage
+
+from .....content import (
+    AssistantContentPart,
+    Text,
+    TextChunk,
+    TextEndChunk,
+    TextStartChunk,
+    Thought,
+    ThoughtChunk,
+    ThoughtEndChunk,
+    ThoughtStartChunk,
+    ToolCall,
+    ToolCallChunk,
+    ToolCallEndChunk,
+    ToolCallStartChunk,
+)
+from .....messages import AssistantMessage
+from .....responses import (
+    AsyncChunkIterator,
+    ChunkIterator,
+    FinishReason,
+    FinishReasonChunk,
+    RawMessageChunk,
+    RawStreamEventChunk,
+)
+from ...model_id import AnthropicModelId, model_name
+
+BETA_FINISH_REASON_MAP = {
+    "max_tokens": FinishReason.MAX_TOKENS,
+    "refusal": FinishReason.REFUSAL,
+}
+
+
+def _decode_beta_assistant_content(content: BetaContentBlock) -> AssistantContentPart:
+    """Convert Beta content block to mirascope AssistantContentPart."""
+    if content.type == "text":
+        return Text(text=content.text)
+    elif content.type == "tool_use":
+        return ToolCall(
+            id=content.id,
+            name=content.name,
+            args=json.dumps(content.input),
+        )
+    elif content.type == "thinking":
+        return Thought(thought=content.thinking)
+    else:
+        raise NotImplementedError(
+            f"Support for beta content type `{content.type}` is not yet implemented."
+        )
+
+
+def decode_response(
+    response: ParsedBetaMessage[Any],
+    model_id: AnthropicModelId,
+) -> tuple[AssistantMessage, FinishReason | None]:
+    """Convert Beta message to mirascope AssistantMessage."""
+    assistant_message = AssistantMessage(
+        content=[_decode_beta_assistant_content(part) for part in response.content],
+        provider_id="anthropic",
+        model_id=model_id,
+        provider_model_name=model_name(model_id),
+        raw_message={
+            "role": response.role,
+            "content": [
+                part.model_dump(exclude_none=True) for part in response.content
+            ],
+        },
+    )
+    stop_reason = response.stop_reason
+    if stop_reason == "model_context_window_exceeded":  # pragma: no cover
+        stop_reason = None
+    finish_reason = BETA_FINISH_REASON_MAP.get(stop_reason) if stop_reason else None
```

Comment:
The "setting stop reason to None if it was model_context_window_exceeded" bit was a hack that I did because I was converting the `ParsedBetaMessage` to an `AnthropicMessage`. Now that we have a proper beta provider we should look it up in BETA_FINISH_REASON_MAP just like any other stop reason. I should have added a comment on this to make it clearer :)